srml/staking/src/lib.rs

```diff
@@ -80,100 +167,267 @@ pub trait Trait: system::Trait + session::Trait {
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		fn bond_extra(origin, max_additional: BalanceOf<T>) {
+			let controller = ensure_signed(origin)?;
+			let mut ledger = Self::ledger(&controller).ok_or("not a controller")?;
+			let stash_balance = T::Currency::free_balance(&ledger.stash);
+
+			if stash_balance > ledger.total {
+				let extra = (stash_balance - ledger.total).min(max_additional);
+				ledger.total += extra;
+				ledger.active += extra;
+				<Ledger<T>>::insert(&controller, ledger);
+			}
+		}
 
-			let target = <Nominating<T>>::get(&source).ok_or("Account must be nominating")?;
+		/// Schedule a portion of the stash to be unlocked ready for transfer out after the bond
+		/// period ends. If this leaves an amount actively bonded less than 
+		/// T::Currency::existential_deposit(), then it is increased to the full amount.
+		/// 
+		/// Once the unlock period is done, you can call `withdraw_unbonded` to actually move
+		/// the funds out of management ready for transfer. 
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		/// 
+		/// See also `withdraw_unbonded`.
+		fn unbond(origin, #[compact] value: BalanceOf<T>) {
+			let controller = ensure_signed(origin)?;
+			let mut ledger = Self::ledger(&controller).ok_or("not a controller")?;
+
+			let mut value = value.min(ledger.active);
+
+			if !value.is_zero() {
+				ledger.active -= value;
+
+				// Avoid there being a dust balance left in the staking system.
+				let ed = T::Currency::minimum_balance();
+				if ledger.active < ed {
+					value += ledger.active;
+					ledger.active = Zero::zero();
+				}
 
-			let mut t = Self::nominators_for(&target);
-			if t.get(target_index) != Some(&source) {
-				return Err("Invalid target index")
+				let era = Self::current_era() + Self::bonding_duration();
+				ledger.unlocking.push(UnlockChunk { value, era });
+				<Ledger<T>>::insert(&controller, ledger);
 			}
+		}
 
-			// Ok - all valid.
-
-			// update nominators_for
-			t.swap_remove(target_index);
-			<NominatorsFor<T>>::insert(&target, t);
+		/// Remove any unlocked chunks from the `unlocking` queue from our management.
+		/// 
+		/// This essentially frees up that balance to be used by the stash account to do
+		/// whatever it wants.
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		/// 
+		/// See also `unbond`.
+		fn withdraw_unbonded(origin) {
+			let controller = ensure_signed(origin)?;
+			let ledger = Self::ledger(&controller).ok_or("not a controller")?;
+			<Ledger<T>>::insert(&controller, ledger.consolidate_unlocked(Self::current_era()));
+		}
 
-			// update nominating
-			<Nominating<T>>::remove(&source);
+		/// Declare the desire to validate for the origin controller.
+		///
+		/// Effects will be felt at the beginning of the next era.
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		fn validate(origin, prefs: ValidatorPrefs<BalanceOf<T>>) {
```

Comment:
I think we should check here that `unstake_threshold` is less than or equal to `MAX_UNSTAKE_THRESHOLD`. We could either not add the item to storage, or modify it on the way in

Reply:
it's already used with `.min(MAX_UNSTAKE_THRESHOLD)` elsewhere in the codebase, but we could also do it on the way in too if desired.

Reply:
I think it makes sense to also do it on the way in, otherwise there could be some bug with someone trying to use the value.

Reply:
Made an update to the logic to return `Err` when user tries to set a value too high